designs/2019-worker-threads/README.md

```diff
@@ -0,0 +1,293 @@
+- Start Date: 2019-09-29
+- RFC PR: https://github.com/eslint/rfcs/pull/42
+- Authors: Toru Nagashima ([@mysticatea](https://github.com/mysticatea))
+
+# Lint files in parallel if many files exist
+
+## Summary
+
+This RFC adds the supports for parallel linting by `worker_threads`/`child_process` modules.
+
+## Motivation
+
+Linting logic uses CPU heavily, so parallel linting by threads may reduce spent time.
+
+## Detailed Design
+
+**PoC**: [eslint/eslint#very-rough-worker-threads-poc](https://github.com/eslint/eslint/tree/very-rough-worker-threads-poc/lib/eslint)
+
+This RFC has two steps.
+
+**The step 1 (semver-minor):**
+
+- Adds `--concurrency` CLI option to specify the number of worker threads. Defaults to `1`.
+- Adds `concurrency` constructor option to `ESLint` class to specify the number of worker threads. Defaults to `1`.
+- Changes the behavior of `ESLint#executeOnFiles(patterns)` method.
+
+**The step 2 (semver-major):**
+
+- Changes the default of the `--concurrency` CLI option to `"auto"`.
+- Changes the default of the `concurrency` constructor option to `"auto"`.
+
+### [STEP 1] New `--concurrency` CLI Option
+
+The `--concurrency` has an argument. The argument should be a positive integer or `"auto"`. Defaults to `1`.
+
+- If `1` is present, ESLint does linting in the main thread.
+- If another integer is present, ESLint creates this number of worker threads then the worker threads do linting.
+
+If `auto` is present, ESLint estimates the best value from the number of target files. It's the number of target files divided by the constant `128`, but `os.cpus().length` at most.
+
+```js
+concurrency = Math.min(os.cpus().length, Math.ceil(targetFiles.length / 128))
+```
+
+This means that ESLint does linting in the main thread if the number of target files is less than `128` in order to avoid the overhead of multithreading. But ESLint does linting with using worker threads automatically if target files are many.
```

Comment:
Can you add some reasoning as to why you opted for `128` as the number here? Without such a reasoning it feels kind of arbitrary 🙂

Reply:
It's my feeling 😄 . While I'm running PoC on my PC, I have seen that it spent a longer time if each worker verified small number of files. I have written a bigger number in this RFC for the case the initialize cost is bigger than my case.